[4/4][ResourceMngmt] Account Bloom/Ribbon filter construction memory in global memory limit

hx235 opened the pull request:
Note: This PR is the 4th part of a bigger PR stack (https://github.com/facebook/rocksdb/pull/9073) and will rebase/merge only after the first three PRs (https://github.com/facebook/rocksdb/pull/9070, https://github.com/facebook/rocksdb/pull/9071, https://github.com/facebook/rocksdb/pull/9130) merge.

**Context:**
Similar to https://github.com/facebook/rocksdb/pull/8428, this PR is to track memory usage during (new) Bloom Filter (i.e,FastLocalBloom) and Ribbon Filter (i.e, Ribbon128) construction by charging dummy entry to block cache, moving toward the goal of [single global memory limit using block cache capacity](https://github.com/facebook/rocksdb/wiki/Projects-Being-Developed#improving-memory-efficiency). It also constrains the size of the banding portion of Ribbon Filter during construction by falling back to Bloom Filter if that banding is, at some point, larger than the available space in the cache under `LRUCacheOptions::strict_capacity_limit=true`.

The option to turn on this feature is `BlockBasedTableOptions::reserve_table_builder_memory = true` which by default is set to `false`. We [decided](https://github.com/facebook/rocksdb/pull/9073#discussion_r741548409) not to have separate option for separate memory user in table building therefore their memory accounting are all bundled under one general option.

**Summary:**
- Reserved/released cache for creation/destruction of three main memory users using CacheReservationManager: 
   - hash entries (i.e`hash_entries`.size(), we bucket-charge hash entries during insertion for performance), 
   - banding (Ribbon Filter only, `bytes_coeff_rows` +`bytes_result_rows` + `bytes_backtrack`), 
   - final filter (i.e, `mutable_buf`'s size).
      - Implementation details: in order to use `CacheReservationManager::CacheReservationHandle` to account final filter's memory, we have to store the `CacheReservationManager` object and `CacheReservationHandle` for final filter in `XXPH3BitsFilterBuilder` as well as  explicitly delete the filter bits builder when done with the final filter in block based table.
- Added option fo run `filter_bench` with this memory reservation feature

**Test:**
- Added new tests in `db_bloom_filter_test` to verify filter construction peak cache reservation under combination of  `BlockBasedTable::Rep::FilterType` (e.g, `kFullFilter`, `kPartitionedFilter`), `BloomFilterPolicy::Mode`(e.g, `kFastLocalBloom`, `kStandard128Ribbon`, `kDeprecatedBlock`) and `BlockBasedTableOptions::reserve_table_builder_memory`
  - To address the concern for slow test: tests with memory reservation under `kFullFilter` + `kStandard128Ribbon` and `kPartitionedFilter` take around **3000 - 6000 ms** and others take around **1500 - 2000 ms**, in total adding **20000 - 25000 ms** to the test suit running locally
- Added new test in `bloom_test` to verify Ribbon Filter fallback on large banding in FullFilter
- Added test in `filter_bench` to verify that this feature does not significantly slow down Bloom/Ribbon Filter construction speed. Local result averaged over **20** run as below:
   - FastLocalBloom 
      - baseline `./filter_bench -impl=2 -quick -runs 20 | grep 'Build avg'`: 
         - **Build avg ns/key: 29.56295** (DEBUG_LEVEL=1), **29.98153** (DEBUG_LEVEL=0)
      - new feature (expected to be similar as above)`./filter_bench -impl=2 -quick -runs 20 -reserve_table_builder_memory=true | grep 'Build avg'`:
         - **Build avg ns/key: 30.99046** (DEBUG_LEVEL=1), **30.48867** (DEBUG_LEVEL=0)
      - new feature of RibbonFilter with fallback  (expected to be similar as above) `./filter_bench -impl=2 -quick -runs 20 -reserve_table_builder_memory=true -strict_capacity_limit=true | grep 'Build avg'` : 
         - **Build avg ns/key: 31.146975** (DEBUG_LEVEL=1), **30.08165** (DEBUG_LEVEL=0)
    
    - Ribbon128
       - baseline `./filter_bench -impl=3 -quick -runs 20 | grep 'Build avg'`:
           - **Build avg ns/key: 129.17585** (DEBUG_LEVEL=1), **130.5225** (DEBUG_LEVEL=0)
       - new feature  (expected to be similar as above) `./filter_bench -impl=3 -quick -runs 20 -reserve_table_builder_memory=true | grep 'Build avg' `:
           - **Build avg ns/key: 131.61645** (DEBUG_LEVEL=1), **132.98075** (DEBUG_LEVEL=0)
       - new feature of RibbonFilter with fallback (expected to be a lot faster than above due to fallback) `./filter_bench -impl=3 -quick -runs 20 -reserve_table_builder_memory=true -strict_capacity_limit=true | grep 'Build avg'` :
          - **Build avg ns/key: 52.032965** (DEBUG_LEVEL=1), **52.597825** (DEBUG_LEVEL=0)
          - And the warning message of `"Cache reservation for Ribbon filter banding failed due to cache full"` is indeed logged to console.

hx235 commented:
Update:
- Rebase on update to https://github.com/facebook/rocksdb/pull/9070

pdillinger commented on include/rocksdb/table.h:
```diff
@@ -342,6 +342,20 @@ struct BlockBasedTableOptions {
+  // (i.e, causing a cache full when strict_capacity_limit = true)
+  //
+  // Default: false
+  bool reserve_bloom_ribbon_filter_construction_memory = false;
```
I prefer simple: one omnibus option for whether to reserve in block cache our best estimate of any non-negligible memory for table construction. I would not assume "everyone" will eventually want to charge everything, because some people use a small block cache because they do not prioritize read performance.

ajkr replied:
> one omnibus option for whether to reserve in block cache our best estimate of any non-negligible memory for table construction

Should such an option include table reader memory too? Or would that be a separate option in `BlockBasedTableOptions`? 

I am inclined toward bundling and not offering ability to disable it for new features. People with small block cache who don't care about reads and have strict_capacity_limit == 0 likely won't notice. Those who will notice have a chance to notice (and adjust their block cache in response) when they read about/try out the new feature, so hopefully we can minimize their chance of only noticing once in production. 

But I don't care that much. Having one option for charging known table reader memory and another option for charging known table builder memory is fine with me.

hx235 replied:
Discussed offline: use `reserve_table_builder_mem`

pdillinger commented on table/block_based/block_based_table_builder.cc:
```diff
@@ -487,7 +488,15 @@ struct BlockBasedTableBuilder::Rep {
       filter_context.column_family_name = tbo.column_family_name;
       filter_context.reason = reason;
-
+      if (table_options.reserve_bloom_ribbon_filter_construction_memory &&
+          !table_options.no_block_cache &&
+          table_options.block_cache != nullptr) {
```
👍  This should be fine.

pdillinger commented on table/block_based/filter_policy.cc:
```diff
@@ -48,10 +50,30 @@ Slice FinishAlwaysFalse(std::unique_ptr<const char[]>* /*buf*/) {
+      std::shared_ptr<CacheReservationManager> cache_res_mgr)
+      : aggregate_rounding_balance_(aggregate_rounding_balance) {
+    cache_res_mgr_ = cache_res_mgr;
+    // Reserve cache for hash_entries_ declaration: sizeof(hash_entries_)
```
`sizeof(hash_entries_)` should be negligible (unimportant) compared to the payload of entries (which you handle below)

pdillinger commented on cache/cache_entry_roles.cc:
```diff
@@ -20,6 +20,7 @@ std::array<const char*, kNumCacheEntryRoles> kCacheEntryRoleToCamelString{{
     "OtherBlock",
     "WriteBuffer",
     "CompressionDictionaryBuildingBuffer",
+    "kXXPH3FilterConstruction",
```
I think just `FilterConstruction` / `kFilterConstruction` is fine.

pdillinger commented on table/block_based/filter_policy.cc:
```diff
@@ -168,10 +199,24 @@ class XXPH3FilterBitsBuilder : public BuiltinFilterBitsBuilder {
       buf->reset(new char[rv]());
+      // Reserve cache for buf
+      if (cache_res_mgr) {
+        Status s = cache_res_mgr->UpdateCacheReservation<
```
You have this pattern repeating all over the place, making for
* A large mass of code for an auxiliary feature
* Difficulty connecting reservation increase point to corresponding reservation decrease point (when applicable?)
* Potential inaccuracy when ignoring status (decrease missing corresponding increase)
* The smell of a threading data race between GetTotalMemoryUsed and UpdateCacheReservation. (I don't think we're dealing with multiple threads here, but it looks suspicious.)

I think it would be easier to understand what was going on if each incremental reservation had its own RAII handle (possibly Cache::Handle) that releases the reservation when released. The FilterBitsBuilder API doesn't make that way easy but I think that's the approach we should be looking at.

hx235 commented:
Update:
- Addressed comments on renaming of `kXXPH3FilterConstruction`, removing negligible memory accounting of `sizeof(hash_entries)`
- Replaced memory accounting of `hash_entries` and `banding` with `CacheReservationHandle`
   - To @pdillinger: as you can see in some commented code (which will be deleted later), I tried using  `CacheReservationHandle` for `mutable_buf` but it seems that we have to add a new parameter `std::unique_ptr<CacheReservationHandle<CacheEntryRole::kFilterConstruction>>` to the public API `FilterBitsBuilder::Finish(buf)`. External customization/inheritance of `FilterBitsBuilder` with no access to non-public header `std::unique_ptr<CacheReservationHandle<CacheEntryRole::kFilterConstruction>>` can't even implement the function. Need to think harder on this :/
- Updated filter bench result, not much difference

hx235 commented:
Update:
- Addressed comments on renaming of `kXXPH3FilterConstruction`, removing negligible memory accounting of `sizeof(hash_entries)`
- Replaced memory accounting of `hash_entries` and `banding` with `CacheReservationHandle`
   - To @pdillinger: as you can see in some commented code (which will be deleted later), I tried using  `CacheReservationHandle` for `mutable_buf` but it seems that we have to add a new parameter `std::unique_ptr<CacheReservationHandle<CacheEntryRole::kFilterConstruction>>` to the public API `FilterBitsBuilder::Finish(buf)`. External customization/inheritance of `FilterBitsBuilder` with no access to non-public header `std::unique_ptr<CacheReservationHandle<CacheEntryRole::kFilterConstruction>>` can't even implement the function. Need to think harder on this :/
- Updated filter bench result, not much difference
- Ready for review when signals are cleared

hx235 commented:
Update:
- not ready for review
- trying something to make using handle for mutable_buf work

hx235 commented:
Update:
- Used `CacheReservationHandle` to manage cache reservation of final filter (`mutable_buf_`), which includes a public API change in `FilterBitsBuilder` and an internal factoring
- Updated filter bench result to be an average of 20 runs
- Updated PR for test information
- Ready for review after signals are cleared

hx235 commented on include/rocksdb/filter_policy.h:
```diff
@@ -65,6 +70,16 @@ class FilterBitsBuilder {
   virtual Slice Finish(std::unique_ptr<const char[]>* buf) = 0;
 
+  // For internal usage
+  virtual Slice Finish(
```
Note to @pdillinger:

I can't think of a different clean way of relaying the reservation handle for final filter (i.e, 'mutable_buf_'), other than having a new `FilterBitsBuilder::Finish` to take it.

For subclasses extending `FilterBitsBuilder` without implementing this new `FilterBitsBuilder::Finish` , when called with this new API, the call will be redirected to use its implemented `FilterBitsBuilder::Finish`, dropping the handle (e.g, `TestFilterBitsBuilder`, `CountUniqueFilterBitsBuilderWrapper ` in [full_filter_block_test.cc](https://github.com/facebook/rocksdb/blob/main/table/block_based/full_filter_block_test.cc)). However, there could be **a compiler warning `[-Werror=overloaded-virtual]`** on those subclasses which requires suppressing or adding `using FilterBitsBuilder::Finish` like I did to them. Not sure if this can be an issue.

For subclasses extending `FilterBitsBuilder` implementing this new `Finish` API (only internal classes `FastLocalBloomBitsBuilder`, `Standard128RibbonBitsBuilder` for now), they can implement the old `Finish` API to call the new API with null handle instead, like I did.

hx235 commented on include/rocksdb/filter_policy.h:
```diff
@@ -65,6 +70,16 @@ class FilterBitsBuilder {
   // The ownership of actual data is set to buf
   virtual Slice Finish(std::unique_ptr<const char[]>* buf) = 0;
 
+  // For internal usage
+  virtual Slice Finish(
+      std::unique_ptr<const char[]>* buf,
+      std::unique_ptr<
+          CacheReservationHandle<CacheEntryRole::kFilterConstruction> >*
+          buf_cache_res_handle) {
+    (void)buf_cache_res_handle;
+    return Finish(buf);
+  }
+
```
Note to @pdillinger:

I can't think of a different clean way of relaying the reservation handle for final filter (i.e, 'mutable_buf_'), other than having a new `FilterBitsBuilder::Finish` to take it. Considering a reservation handle might want to stay as close to its filter data as possible.

For subclasses extending `FilterBitsBuilder` without implementing this new `FilterBitsBuilder::Finish` , when called with this new API, the call will be redirected to use its implemented `FilterBitsBuilder::Finish`, dropping the handle (e.g, `TestFilterBitsBuilder`, `CountUniqueFilterBitsBuilderWrapper ` in [full_filter_block_test.cc](https://github.com/facebook/rocksdb/blob/main/table/block_based/full_filter_block_test.cc)). However, there could be **a compiler warning `[-Werror=overloaded-virtual]`** on those subclasses which requires suppressing or adding `using FilterBitsBuilder::Finish` like I did to them. Not sure if this can be an issue.

For subclasses extending `FilterBitsBuilder` implementing this new `Finish` API (only internal classes `FastLocalBloomBitsBuilder`, `Standard128RibbonBitsBuilder` for now), they can implement the old `Finish` API to call the new API with null handle instead, like I did.

pdillinger replied:
From chat: I think we can just release the handle on destroying the FilterBitsBuilder, and within RocksDB do something to tell the FilterBlockBuilder to destroy its FilterBitsBuilder.

hx235 commented on table/block_based/filter_policy.cc:
```diff
@@ -207,12 +215,47 @@ class FastLocalBloomBitsBuilder : public XXPH3FilterBitsBuilder {
   virtual Slice Finish(std::unique_ptr<const char[]>* buf) override {
+    return Finish(buf, nullptr);
+  }
+  // REQUIRED: if cache_res_mgr_ != nullptr, then buf_cache_res_handle !=
+  // nullptr
```
Note to @pdillinger:

I think it makes sense to have this assertion. Otherwise we might end up in situation where there is actual memory reservation but it will be released way earlier than the final filter get de-allocated since there is no handle to rely that reservation.

This PR also makes sure all path calling here with `cache_res_mgr_ != nullptr` (i.e, only `BlockBasedTableBuilder` and `FilterBench`) provide a non-null handle.

hx235 replied:
No longer needed due to API change

hx235 commented on table/block_based/filter_block.h:
```diff
@@ -75,12 +78,28 @@ class FilterBlockBuilder {
+  // ownership of cache reservation for the memory used by filter data
+  // BlockBasedFilterBlock will ignore this parameter.
+  //
+  // REQUIRED:
+  // If BlockBasedTableOptions:: reserve_table_builder_memory == true,
+  // and a block cache is availabe, that is, we charge filter construction
+  // memory, then we need the following to be true for
+  // filter_data_cache_res_handle to manage filter_data's cache reservation
+  // correctly:
+  // (filter_data && filter_data_cache_res_handle) || (!filter_data &&
+  // !filter_data_cache_res_handle)
```
Note to @pdillinger :
We allows either `filter_data` or `filter_data_cache_res_handle` to be `nullptr` for compatibility with existing usage of `FilterBlock::Finish()`. But we have a stricter constrain on them when memory reservation is needed. 

I've been thinking of a way to write an assertion for this requirement but find it hard to access information of whether we want memory reservation or not at `FilterBlock` level.

Maybe we can do `FilterBlock::Finish(tmp, filter_data = nullptr, reserve_filter_construction_mem = false, filter_data_cache_res_handle = nullptr)` and assert the following:
`
if (reserve_filter_construction_mem) {
    assert (filter_data && filter_data_cache_res_handle) || (!filter_data && !filter_data_cache_res_handle) 
}
`

hx235 replied:
No longer needed due to API change

hx235 commented on table/block_based/filter_block.h:
```diff
@@ -75,12 +78,28 @@ class FilterBlockBuilder {
+  // possible. BlockBasedFilterBlock will ignore this parameter.
+  //
+  // If filter_data_cache_res_handle is null, Finish() may transfer
+  // ownership of cache reservation for the memory used by filter data
```
Todo:
“.. ownership of cache reservation when reserving feature is available …”

hx235 replied:
No longer needed due to API change

hx235 commented on include/rocksdb/table.h:
```diff
@@ -490,6 +490,25 @@ struct BlockBasedTableOptions {
+  // at that point (i.e, causing a cache full when strict_capacity_limit =
+  // true), construction will fall back to (new) Bloom Filter.
+  //
+  // Default: false
```
Todo:
Move the option closer to block cache related option and explicit state that block cache needs to be available for this option to take effect

hx235 replied:
Done

pdillinger commented on include/rocksdb/filter_policy.h:
```diff
@@ -139,6 +154,10 @@ struct FilterBuildingContext {
+
+  // Optional: if provided, it charges memory usage of bloom filter/ribbon
+  // filter construction to block cache
+  std::shared_ptr<CacheReservationManager> cache_res_mgr;
```
From chat: The FilterBitsBuilder should be able to "self serve" by looking up options in the `table_options`. Thus, I don't think we need to expose CacheReservationManager

hx235 commented:
Update:
- Not ready for review
- Just go through CI test

hx235 commented:
Update:
- Addressed feedback on API comments, self-service CRM within FilterBitsBuilder, explicit reset FilterBitsBuilder after done with the final filter in block based table builder
- Some minor improvement
- Updated filter benchmarking result (not much difference from previous result)
- Ready for update when signals are cleared

pdillinger commented on table/table_test.cc:
```diff
@@ -5367,6 +5369,181 @@ TEST_P(
+  }
+
+  ASSERT_OK(builder->Finish());
+  EXPECT_EQ(cache->GetPinnedUsage(), 0 * kSizeDummyEntry);
```
I don't see this test checking that memory is actually reserved. What is the intent?

I think we should use a sync point callback at the point of expected peak reservation, or a custom Cache wrapper, to ensure the memory reservation reaches something around what is expected. IMHO, detecting the peak is a sufficient regression test. I'm willing to believe the "area under the curve" is accounted over time by inspection.

pdillinger commented on util/bloom_test.cc:
```diff
@@ -601,6 +603,255 @@ TEST_P(FullBloomTest, OptimizeForMemory) {
+  }
+}
+
+class CustomFilterPolicy : public FilterPolicy {
```
Not sure why this class is copy-pasted. IMHO this PR doesn't do anything to make it susceptible to failure under a wrapper custom filter policy but not under a built-in filter policy, so this custom filter policy seems like extra, unnecessary test code (that we have to maintain forever).

pdillinger commented on util/bloom_test.cc:
```diff
@@ -601,6 +603,255 @@ TEST_P(FullBloomTest, OptimizeForMemory) {
+  EXPECT_EQ(cache->GetPinnedUsage(), 0 * kSizeDummyEntry);
+}
+
+TEST_P(FullBloomTest, RibbonFilterFallBackOnLargeBanding) {
```
I agree this fallback case is worth testing. I'm ok with this one being tested at the unit (FilterBitsBuilder) level because it is a little trickier (but doable) at the integration level, and other tests should give us confidence in the integration logic (which this fallback case does not complicate).

pdillinger commented on util/bloom_test.cc:
```diff
@@ -601,6 +603,255 @@ TEST_P(FullBloomTest, OptimizeForMemory) {
   }
 }
 
+TEST_P(FullBloomTest, ReserveBloomAndRibbonFilterConstructionMemory) {
```
This seems like too much unit test for the job we are doing here. IMHO, regressions are at least as likely in the integration logic as they are in the code units tested here. Unit tests at this level of detail make code evolution substantially more difficult, so have to offer sufficient value to justify their existence. While this is an important feature, it is among the least "critical" that we write.

IMHO we can reasonably cover this feature with an integration test (as in table_test; see other comment on testing), and partitioned filter should just change the calculation on peak usage.

pdillinger commented on util/bloom_test.cc:
```diff
@@ -601,6 +602,351 @@ TEST_P(FullBloomTest, OptimizeForMemory) {
+  const std::size_t num_entries = 10000;
+  char key_buffer[sizeof(int)];
+  for (std::size_t i = 0; i < num_entries; ++i) {
+    filter_bits_builder->AddKey(Key(static_cast<int>(i), key_buffer));
```
If you have even a rough bound on peak memory usage by Bloom construction, that suffices to check for hash_entries because that is the overwhelming majority of usage for Bloom.

pdillinger commented on util/bloom_test.cc:
```diff
@@ -601,6 +603,255 @@ TEST_P(FullBloomTest, OptimizeForMemory) {
+      filter_bits_builder->AddKey(Key(static_cast<int>(i), key_buffer));
+    }
+    // 0xffffffffU is max size supported by implementation for buf length
+    std::unique_ptr<const char[]> buf(new char[0xffffffffU]);
```
We should not be allocating a 4GB memory buffer in unit tests. And I don't think there's any reason to allocate something to `buf` here.

pdillinger commented on util/bloom_test.cc:
```diff
@@ -601,6 +602,351 @@ TEST_P(FullBloomTest, OptimizeForMemory) {
+    // -1 = Marker for newer Bloom implementations
+    // -2 = Marker for Standard128 Ribbon
+    if (will_fall_back) {
+      EXPECT_EQ(filter.data()[filter.size() - 5], static_cast<char>(-1));
```
I'm satisfied with your single control (`!will_fall_back case`) but if you wanted to also control your other concern, you could have another case/iteration that turns off `reserve_table_builder_memory`.

pdillinger commented on util/bloom_test.cc:
```diff
@@ -601,6 +603,255 @@ TEST_P(FullBloomTest, OptimizeForMemory) {
+  // Before banding: hash_entries_.size() * hash_entries_[0] ≈ 50,000 * 8 =
+  // 400,000 < 0.4MB Banding'size ≈ 1,034,288 < 2MB After banding: 0.4 MB + 2MB
+  // < 8MB
+  for (std::size_t num_entries : {500000, 50000}) {
```
It should not be necessary to use large, expensive constructions in unit tests. You should be able to scale the parameters down to something super cheap and fast. (We are not introducing new reasons to be concerned with e.g. value overflow.)

pdillinger commented on util/bloom_test.cc:
```diff
@@ -601,6 +603,255 @@ TEST_P(FullBloomTest, OptimizeForMemory) {
+  const std::unique_ptr<const FilterPolicy> policy_otherwise_;
+};
+
+TEST(FullBloomCustomFilterPolicyTest, RibbonFilterFallBackOnLargeBanding) {
```
More copy-paste

pdillinger commented on util/bloom_test.cc:
```diff
@@ -601,6 +603,255 @@ TEST_P(FullBloomTest, OptimizeForMemory) {
+    // filter.size())
+    memory_reserved = filter.size();
+    num_dummy_entries = static_cast<std::size_t>(
+        std::ceil(memory_reserved / static_cast<double>(kSizeDummyEntry)));
```
Btw, I usually use `1.0 * ` to coerce something to double without a cast

hx235 commented on table/block_based/filter_policy.cc:
```diff
@@ -209,10 +224,36 @@ class FastLocalBloomBitsBuilder : public XXPH3FilterBitsBuilder {
+        (num_entries == 0 ? 0 : num_entries * sizeof(hash_entries_[0]));
+
+    // Cache reservation for hash entries
+    std::unique_ptr<CacheReservationHandle<CacheEntryRole::kFilterConstruction>>
```
Discussed offline: 
(1) try charge memory of a bucket of keys instead of per key. sizeof(bucket) = dummy entry size/sizeof (key hash)
(2) check the performance

hx235 commented on table/table_test.cc:
```diff
@@ -5367,6 +5369,181 @@ TEST_P(
+
+INSTANTIATE_TEST_CASE_P(
+    BlockBasedTableOptions, BlockBasedTableReserveFilterConstructionMemoryTest,
+    ::testing::Values(
```
Discussed offline: integration in RocksDB usually uses public API to verify behavior and user experience

pdillinger replied:
All combinations: https://stackoverflow.com/a/13240573/454544

hx235 commented:
Update
- Addressed comments on charging hash entry bucket, peak testing and coding
- Updated PR and filter bench results
- Ready for review after signals are cleared

hx235 commented on db/db_bloom_filter_test.cc:
```diff
@@ -667,6 +670,400 @@ TEST_F(DBBloomFilterTest, BloomFilterReverseCompatibility) {
+      // two partitions.
+      num_key_ = 18 * CacheReservationManager::GetDummyEntrySize() /
+                 sizeof(FilterConstructionReserveMemoryHash);
+    } else if (policy_ == BloomFilterPolicy::Mode::kFastLocalBloom) {
```
Note: in this case, I think we can reduce number of hash keys to less than 6 dummy entries in verifying we indeed charge final filter as explained by the comment.

hx235 commented on db/db_bloom_filter_test.cc:
```diff
@@ -667,6 +670,400 @@ TEST_F(DBBloomFilterTest, BloomFilterReverseCompatibility) {
+                predicted_filter_construction_cache_res_peak * 1.1);
+      return;
+    } else {
+      EXPECT_GE(filter_construction_cache_res_peaks.size(), 2)
+          << "Filter construction cache reservation should have multiple peaks "
+             "in case: BloomFilterPolicy::Mode::kFastLocalBloom + "
+             "PartitionedFilter";
```
Note: for some reason, this assertion fails in one of the CI test run : https://app.circleci.com/pipelines/github/facebook/rocksdb/9993/workflows/1bbb4523-0be6-4550-8015-1f344467d7fb/jobs/239075 but not locally. And I can't reproduce that on the CI machine or locally.

Is there any potential reason why number of keys per partition can be different among different environment?

pdillinger replied:
I suspect the final filters are not large enough to cause a second peak. If you jack up metadata_block_size to be like reservation increment size, then it should work (though we cut a partition boundary with either index or filter data fills to that block size; so not a guarantee if index size is dominating).

hx235 commented on db/db_bloom_filter_test.cc:
```diff
@@ -667,6 +670,400 @@ TEST_F(DBBloomFilterTest, BloomFilterReverseCompatibility) {
+                Handle** handle = nullptr,
+                Priority priority = Priority::LOW) override {
+    Status s = target_->Insert(key, value, charge, deleter, handle, priority);
+    auto map = CopyCacheDeleterRoleMap();
```
Note: reading `CopyCacheDeleterRoleMap()` API leaves me with impression that I have to call this function every time a deleter is registered, which can be another overhead for this test.

pdillinger replied:
You can just compare to `GetNoopDeleterForRole<CacheEntryRole::kFilterConstruction>()` and for extra speed, save that ahead of time to a static variable or Test object field.

hx235 commented on table/block_based/filter_policy.cc:
```diff
@@ -61,6 +66,20 @@ class XXPH3FilterBitsBuilder : public BuiltinFilterBitsBuilder {
+                    kUint64tHashEntryCacheResBucketSize * sizeof(hash),
+                    &hash_entry_cache_res_bucket_handles_.back());
+        s.PermitUncheckedError();
+      }
```
Note: compared with previous solution, bucket charging causes around +1.5ns/key for Bloom Filter and +2.5ns/key for Ribbon Filter. Should be fine (?)

pdillinger replied:
Seems fine. Thanks

hx235 commented:
Update
- Addressed comment on memory accounting gap, testing peak and coding suggestion
- Updated benchmarking result and PR, left some notes along the PR
- Ready for review after signals are cleared

hx235 commented:
Update
- Addressed comment on memory accounting gap, testing peak and coding suggestion
- Updated benchmarking result and PR, left some notes along the PR
- Ready for review after signals are cleared

pdillinger commented on table/block_based/filter_policy.cc:
```diff
@@ -61,6 +66,20 @@ class XXPH3FilterBitsBuilder : public BuiltinFilterBitsBuilder {
     if (hash_entries_.empty() || hash != hash_entries_.back()) {
       hash_entries_.push_back(hash);
+      if (cache_res_mgr_ &&
+          (hash_entries_.size() % (kUint64tHashEntryCacheResBucketSize / 2)) ==
```
I think we can simplify:

    if (cache_res_mgr_ &&
        // Traditional rounding to whole bucket size
        (hash_entries_.size() % kUint64tHashEntryCacheResBucketSize) == kUint64tHashEntryCacheResBucketSize / 2) {

pdillinger commented on table/block_based/filter_policy.cc:
```diff
@@ -1135,6 +1234,14 @@ FilterBitsBuilder* BloomFilterPolicy::GetBuilderWithContext(
+  bool reserve_filter_construction_mem =
+      (context.table_options.reserve_table_builder_memory &&
+       context.table_options.block_cache);
+  std::shared_ptr<CacheReservationManager> cache_res_mgr;
```
Considering that we can create CacheReservationManagers on the fly, it might make more sense to pass this one down as unique_ptr, so that it does not incur memory synchronization overhead associated with shared_ptr (for refcounting each time it's copied). Is that too much of a pain here?

hx235 replied:
I thought of that but the issue is CacheReservationManagerHandle needs to share ownership to CacheReservationManager due to std::enable_shared_from_this<T>::shared_from_this by pointing to it.

pdillinger replied:
Oh right. The users of the Handles could pinky swear to close them before the CRM, but that's not a safe, friendly API.

pdillinger commented on include/rocksdb/table.h:
```diff
@@ -285,6 +285,25 @@ struct BlockBasedTableOptions {
   // separately
   uint64_t metadata_block_size = 4096;
 
+  // If true, an estimated memory usage of table building
```
"an estimated memory usage" - As stated, it's not totally clear that we have a dynamically updating charge loosely based on actual usage.

pdillinger commented on db/db_bloom_filter_test.cc:
```diff
@@ -667,6 +670,401 @@ TEST_F(DBBloomFilterTest, BloomFilterReverseCompatibility) {
+  std::shared_ptr<FilterConstructResPeakTrackingCache> cache_;
+};
+
+std::vector<bool> reserve_builder_memory_options{true, false};
+std::vector<BloomFilterPolicy::Mode> policies{
+    BloomFilterPolicy::Mode::kFastLocalBloom,
+    BloomFilterPolicy::Mode::kStandard128Ribbon,
+    BloomFilterPolicy::Mode::kDeprecatedBlock};
+std::vector<bool> partition_filters_options{false, true};
```
In instantiating test cases, I advocate an approach like how MC/DC code coverage[*] tackles the exponential blow-up problem: for each parameter, there should be two instantiations differing in only that parameter. Obviously you want to cover the range of values, but there doesn't have to be blow-up e.g. in mostly useless reserve_builder_memory_options=false instantiations.

I suggest
{false, kFastLocalBloom, false}
{true, kFastLocalBloom, false}
{true, kFastLocalBloom, true}
{true, kStandard128Ribbon, false}
{true, kStandard128Ribbon, true}
{true, kDeprecatedBlock, false}
{true, kLegacyBloom, false}

(Yes, before I was critical of listing a bunch of instantiations, but if you trim them down strategically, I think it's better to save the testing runtime. We have some terribly slow, IMHO poorly planned tests in our suite with hundreds of instantiations.)

[*] https://en.wikipedia.org/wiki/Modified_condition/decision_coverage

hx235 commented:
Update
- Addressed feedback on test instantiation, traditional rounding to whole bucket size, comparison to static noop cache delete function, api documentation
- Disabled auto compaction to prevent its side effect to the test as described in 
- Fixed a typo in test variable naming
- Updated testing time in PR
- No update to the filter benchmarking time as the change above does not and should not change the performance
- Ready for review after signals are cleared

hx235 commented:
Update:
- Added a translation unit hack

pdillinger commented on options/options_settable_test.cc:
```diff
@@ -175,6 +175,7 @@ TEST_F(OptionsSettableTest, BlockBasedTableOptionsAllFieldsSettable) {
       "optimize_filters_for_memory=true;"
       "index_block_restart_interval=4;"
       "filter_policy=bloomfilter:4:true;whole_key_filtering=1;"
+      "reserve_table_builder_memory=false;"
```
Ideally you should also add a small case to OptionsTest.GetBlockBasedTableOptionsFromString to ensure configuration string is properly connected to the field

pdillinger commented on db/db_bloom_filter_test.cc:
```diff
@@ -667,6 +670,411 @@ TEST_F(DBBloomFilterTest, BloomFilterReverseCompatibility) {
+        std::make_tuple(true, BloomFilterPolicy::Mode::kDeprecatedBlock, false),
+        std::make_tuple(true, BloomFilterPolicy::Mode::kLegacyBloom, false)));
+
+TEST_P(DBFilterConstructionReserveMemoryTestWithParam, ReserveMemory) {
```
Maybe drop a TODO comment here about improving running time of the test, possible approach & challenges that presents to keeping the production code simple and the test reflecting production reality.

pdillinger reviewed (approved):
LGTM, thanks!

hx235 commented:
Update:
- Addressed comment on option test and todo comment
- Rebased
- Will merge after signals are cleared
